s2i/builder.go

```diff
@@ -0,0 +1,115 @@
+package s2i
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"os"
+
+	dockerClient "github.com/docker/docker/client"
+
+	"github.com/openshift/source-to-image/pkg/api"
+	"github.com/openshift/source-to-image/pkg/api/validation"
+	"github.com/openshift/source-to-image/pkg/build"
+	"github.com/openshift/source-to-image/pkg/build/strategies"
+	"github.com/openshift/source-to-image/pkg/scm/git"
+
+	fn "knative.dev/kn-plugin-func"
+	docker "knative.dev/kn-plugin-func/docker"
+)
+
+var (
+	// ErrRuntimeRequired indicates the required value of Function Runtime was not provided
+	ErrRuntimeRequired = errors.New("runtime is required to build")
+
+	// ErrRuntimeNotSupported indicates the given runtime is not (yet) supported
+	// by this builder.
+	ErrRuntimeNotSupported = errors.New("runtime not supported")
+)
+
+// DefaultBuilderImages for s2i builders indexed by Runtime Language
+var DefaultBuilderImages = map[string]string{
+	"node": "registry.access.redhat.com/ubi8/nodejs-16", // TODO: finalize choice and include version
+}
+
+// Builder of Functions using the s2i subsystem.
+type Builder struct {
+	verbose bool
+}
+
+// NewBuilder creates a new instance of a Builder with static defaults.
+func NewBuilder(verbose bool) *Builder {
+	return &Builder{verbose: verbose}
+}
+
+func (b *Builder) Build(ctx context.Context, f fn.Function) (err error) {
+	// Ensure the Function has a builder specified
+	if f.Builder == "" {
+		f.Builder, err = defaultBuilderImage(f)
+		if err != nil {
+			return
+		}
+	}
+
+	client, endpoint, err := docker.NewClient(dockerClient.DefaultDockerHost)
```

Comment:
The returned endpoint is docker host from running container PoV (as needed by Buildpack). I believe it shouldn't be used in this context here.

Reply:
The first line of `docker.NewClient` is `os.Getenv("DOCKER_HOST")`.  How is that not the host we want to use?  I am confused and that function contains no documentation.

Reply:
It reads the `DOCKER_HOST` envvar but it may or may not return it.
* For local connection (unix socket and windows named pipe) it returns the `DOCKER_HOST` directly.
* For `ssh` connection it reads the `DOCKER_HOST` from the `ssh` remote.
* For `TCP` connections it returns `""` so it defaults in the remote (and no you cannot use `client.DefaultDockerHost`), this is needed beaus of `TCP+tls` connections.

Reply:
>The returned endpoint is docker host from running container PoV (as needed by Buildpack). 

This is not accurate. It's docker connection string as seen by the machine running the daemon.
So there is some special handling for `ssh` and `TCP` connections.

Reply:
> and no you cannot use `client.DefaultDockerHost`
Because the value is different for various platforms: if you run `func` on Windows and connected via `TCP` to Linux there would be mismatch ("//./pipe/docker_engine" vs "unix:///var/run/docker.sock").